### ISSUE TITLE
SASGRID - Job is remaining in Queue afer a Disconnect

### ISSUE THREAD

**Issue by JLetrange** [sassoftware/saspy#221]
Currently we are making some testings with SASpy to enable connections to a SAS Grid. We are facing on some issues during the disconnection from SAS Grid. The job created on the session establishment is not killed and remain in the queue (displayed through command bjobs). We made some analysis inside the java code and changed/recompiled the library to try to isolate the issue. It seems that the problem is located on the call to the method EnableDisconnect. By removing the call to this method we were able to kill properly the job from the Queue with a simple disconnect from Python side. No documentation was found on our side about this method EnableDisconnect, if there is another way to call it, with other values as input arguments. 

A similar issue was already raised: #131

**Comment by tomweber-sas** [comment]
Can you tell me exactly what you are doing to try to end the job? Since you're using the word disconnect a lot, I'm thinking you are using the disconnect() method. 

If that is the case, which it sounds like. it's doing what it's supposed to do. The disconnect() method allows you to disconnect from the workspace server you're connected to and then reconnect back to that same server later, before the timeout period it's configured to have expires. It does not terminate the SAS session. Simply ending the python process saspy is running in will end the SAS session. You can also explicitly end SAS by using the _endsas() method. Or you can execute the python del() method on the SASsession object; del(sas). 

If you are using disconnect(), that is why the workspace server isn't going away; that's what it's for.

Let me know if that's what is going on or if these are somehow not terminating when you terminate python or issue _endsas() yourself. Also remember, if you're using Jupyter notebooks, that just closing the browser window does not shut down the notebook process. You also have to explicitly stop the notebokk process, and, when the process is stopped, saspy will terminate the SAS session it's connected to.

Thanks,
Tom

**Comment by JLetrange** [comment]
Hi Tom thanks for your quick feedback :)

From python side no command _disconnect, endsas, del(sas)_ ends properly the job on SASGrid:
- sample code used:
```
import saspy

sas = saspy.SASsession(cfgname='winiomwin2')`
res = sas.submit("data xxx; myval=0; run;")`

#sas.disconnect()
#sas._endsas() 
del(sas)`
```

When the command disconnect is called from python we arrive on Java side to the execution of the following code: 

     else if (eol.contains("DISCONNECT"))
         {
         if (reconnect)
            {
            cx.close();
            errp.write("Succesfully disconnected. Be sure to have a valid network connection before submitting anything else.DISCONNECT");
            }
         else
            errp.write("This workspace server is not configured for reconnecting. Did not disconnect.DISCONNECT");

         errp.flush();
         pgm = pgm.substring(idx + 13 + 33);

         }
 
Reconnect is authorized, and so cx.close() is called. At this step should the job be killed or should it remain on the queue ? From my understanding the call to cx.close should end and kill the job. 

In the connect method (Java side) if we replace the code:

      try
          {
          uriStr    = iDisco1.EnableDisconnect(0,false);
          uri       = SASURI.create(uriStr);
          reconnect = true;
          }
       catch (iomReconnectNotAllowed | iomReconnectInvalidTimeout | iomReconnectDisabled |
                     iomEnableFailed | iomNoReconnectPortsAvailable | GenericError e1)
          {
          reconnect = false;
          }

 with only:
 `     reconnect = true; /* only for testing purpose - uri not determined here */
 `
 
the job is properly killed and removed from the queue after the cx.close() call.

**Comment by tomweber-sas** [comment]
Hey, thanks for showing some code. What version of saspy are you running? Again, disconnect() will not shut down the workspace server, _endsas() will. I see that python's del() methods does not necessarily drive the del() method of the object you pass it (the SASsession object), I assume due to some deferred garbage collection algorithm? So del(sas) is not a reliable way to try to shut down either. It may or may not drive the code path right away. Terminating the python process will though.

Can you tell me what code you were actually running when you saw this behavior; before you started changing the java code? Were you using disconnect()?

What I would like to see is going back to the real saspyiom.jar and using _endsas() method and seeing that it terminates the workspace server. Only in the first version where I implemented disconnect did I miss the case where I had to do the extra work to really terminate the session. And I fixed that way back then. Have never heard of an issue since then, and that fixed the problem then. So, I'm a little curious about what's going on here. 

If necessary, we can do a web-ex tomorrow afternoon, if that' s convenient, and we can get to the bottom of this. It's always possible there's some different issue I haven't seen before. And if so, that would get me the info I need much quicker. Of course, if that's not feasible, we can work through here too. No problem.

So, what version of saspy are you running? Put the original saspyiom.jar back in place. Use sas._endsas() and let's see what we get. That's the first thing to see.

Thanks!
Tom

**Comment by JLetrange** [comment]
Hi Tom,

I get the last version from SASPy (version 2.4.3) used the command sas._endsas() and get the same result, the job remains in the queue.

It will more convenient to do a demo in live, I will organize a call to discuss with you directly :) You will receive an invitation on your private email.

Regards,
Joel.

**Comment by tomweber-sas** [comment]
@JLetrange I have been investigating this since our web-ex session. I believe I have a fix that addresses your case, without needing an option at all. There is a disable-reconnect call I can make, and then the one and only close will terminate with no need to reconnect and close again. This is at master now.

I've been trying to get access to a grid running LSF here, so I could try to reproduce what you are seeing. I've been able to get at one, and run it in 2 other ways than having LSF start the workspace servers (the jobs you see with bjobs), but not with LSF starting them; some configuration error no one can track down yet. So, I still haven't been able to see if I see the same behavior.

I'll continue to try to get this working here so I can see about whether the reconnect is really creating a new server or not. There is something you can run for me to tell if that is the case or not, on your end. 
Can you run the following code and see if the 2 tables (A and B) show up in the output from the second list_tables()? It's ok if there's other stuff in work, so long as those two table that were created are still there. That way we can tell if you were reconnected back to the same workspace server or not.

In all of my cases I am reconnecting, but as I haven't been able to test the LSF launched scenario, and we couldn't see the actual workspace server process on your system, this test will tell us if you are reconnecting or creating a new server.

```
import saspy
sas = saspy.SASsession()
sas
sas.submit('data a;x=1;run;','text')
sas.submit('data b;x=1;run;','text')
sas.list_tables('work', 'pandas')
sas.disconnect()
sas.list_tables('work', 'pandas')
```

The fix for endsas() is at master. Can you pull that and verify it? If you need help doing that just let me know. If that fixes your case, then I'll go ahead and promote it and the rest of the changes at master to a new release.

Thanks!
Tom

**Comment by JLetrange** [comment]
Hi Tom,

thanks for all the research you made on SASGrid. I run the piece of code you mentioned with the current version, and with an updated package including the last changes brought.

**1. Run with current version 2.4.3** 

by running the code with the current version 2.4.3 I get the following result:

```
Out[1]: 
  MEMNAME MEMTYPE
0       A    DATA
1       B    DATA

sas.disconnect()
Out[2]: 'Succesfully disconnected. Be sure to have a valid network connection before submitting anything else.'

sas.list_tables('work', 'pandas')
Traceback (most recent call last):

  File "<ipython-input-3-2d0f5cc14a06>", line 1, in <module>
    sas.list_tables('work', 'pandas')

  File "D:\CDP_QSTR\anaconda3\lib\site-packages\saspy\sasbase.py", line 1297, in list_tables
    exists = ll['LOG'].rsplit('LIBREF_EXISTS=')[2].split('\n')[0]

IndexError: list index out of range

```

**2. Run with updated package**

After having changed the files sasbase.py and saspyiom.jar I get the following result:

```
Out[2]: 
  MEMNAME MEMTYPE
0       A    DATA
1       B    DATA

sas.disconnect()
Out[3]: 'Succesfully disconnected. Be sure to have a valid network connection before submitting anything else.'

sas.list_tables('work', 'pandas')
Libref provided is not assigned
```

Before the call to disconnect I get one job in the queue. After the call to disconnect the job still appear in job queue. BUT after the call to sas.list_tables **two additional jobs were added**.


**3. Additional testings**

I checked the fix for _endsas, now the job is no more remaining in the job queue in case only _endsas is called. If the function disconnect is called just before the function _endsas, the job remain in the queue. 

Sample code below:

```
import saspy
sas = saspy.SASsession(cfgname='winiomwin')
sas.submit('data a;x=1;run;','text')
sas._endsas() 
```
==> Job removed from queue

```
import saspy
sas = saspy.SASsession(cfgname='winiomwin')
sas.submit('data a;x=1;run;','text')
sas.disconnect()
sas._endsas() 
```
==> **Job is remaining in the queue**

Do not hesitate to recontact me if you need additional testings.

Regards,
Joel.

**Comment by tomweber-sas** [comment]
That's all a bit strange. I see that the error in case 1 is a bug that is fixed at master; about parsing the log and getting the index out of range. 
Case 2 I'm not sure about. Did you pull master or only update 2 of the files from there? There are changes there to other files as well, so I'm not sure if there's a mismatch with that.
Case 3 implies no errors, just that now the endsas() case is working correctly for you (which is the case I believe should), but w/ disconnect you get a second job, which goes away on endsas, but the first job stays around, like originally reported?

Without LSF being the one to launch the workspace servers, (IOM is doing it), all of these cases work as expected for me. I still don't have a working Grid with LSF launched servers to be able to see bjobs jobs or to see if the interaction of IOM and LSF isn't working right for the disconnect case - as you are seeing.
I will continue to try to get access to a system here where I can test this out and see what is happening.

Is it correct that, with all the code at master, your case where you do not use disconnect(), and you simply connect, do work, and terminate, is now behaving correctly? Only one job created and that job goes away when you do endsas or just close the python process (that implicitly drives endsas under the covers)? That is what I believe you show in case 3.

I will build a new release from master and push that to pypi. that way there will be a fresh clean install you can get which should address the other issues and leave you in a good working state, other than the disconnect problems which still seem outstanding. But you're not using that and I'll continue to track that down.

Thanks!
Tom

**Comment by JLetrange** [comment]
Hi Tom,

The test related to Case 1 was based on an original 2.4.3 version - Package was removed and reinstalled from the repository.

For test 2 I changed only the files sasbase.py and saspyiom.jar and not all the files. I see that meanwhile a new package 2.4.4 was pushed. New testings were made on this version. Like before I get a 'Libref provided is not assigned' after the disconnect.

`Case 3 implies no errors, just that now the endsas() case is working correctly for you (which is the case I believe should), but w/ disconnect you get a second job, which goes away on endsas, but the first job stays around, like originally reported?`

I confirm that calling disconnect just before _endsas results with the same initial job (same job Id) remaining in the queue. If no call is made to disconnect everything is working as expected.

Regards,
Joel.

**Comment by tomweber-sas** [comment]
Thank you @JLetrange !  There is something clearly not behaving right with disconnect when the Grid Workspace servers are LSF launched. That is the only case I can't run, as of yet, here. I'm glad you grabbend 2.2.4 and that it is, at least, addressing your normal use case. I will continue to see about getting an environment to test this with here.  For the sake of maybe seeing some more info, you might just try the following and see what it shows us.

```
import saspy
sas = saspy.SASsession(cfgname='winiomwin')
sas.submit('data a;x=1;run;','text')
print(sas.saslog())
sas.disconnect()
sas.submit('data b;x=1;run;','text')
print(sas.saslog())
sas._endsas() 
```

Looking at the two logs printed out would show us if they are the same server or not or if we even have a connection after trying to reconnect after the disconnect.

Thanks!
Tom

**Comment by JLetrange** [comment]
Hi Tom,

I will provide you the log directly through your email

Regards,
Joel.

**Comment by tomweber-sas** [comment]
Thank you. I received them, and clearly, they are two different SAS sessions, not a re-connection to the first one. This is not what I see when LSF is not the one starting them. When IOM is the one starting them, reconnect gets back to the first, and only, session.
That matches what I was told, that if LSF shows 2 jobs, there are 2 sessions. So this looks like a problem with reconnect in IOM when using LSF. I will continue to track this down here.

Thanks you for confirming this for me!
Tom

**Comment by tomweber-sas** [comment]
@JLetrange I finally got a SAS Grid with LSF working here where LSF is the one launching the workspace server. It's only a single node configuration, and I don't know if that makes a difference (maybe). But, it worked as it should. I reconnect to the one and only workspace server with only the one LSF job ever being started.
I've asked if we have a multi-node setup that I can try out, but I don't know if there is one or not at this point. 
I think that at this point I'd like to have you open a tech support track for this problem. It is clearly the wrong behavior, but as I see it work correctly in all of the cases I've been able to try, and we're running the same saspy code, doing the same (and only) IOM calls, there's something else that is causing this discrepancy. And I have no expertise in Grid and LSF configuration or the underlying IOM code itself that is also integrated with this. I think driving this through tech support will be the best case for getting this resolved. I suspect the first thing they will request if the Object Spawner log, with trace level set to Debugl, as opposed to Info, which is likely the default.
Can you do that? assuming so, let me know the track number so I can get on it and provide all the info I have and make sure the right people get on it right away,
Thanks!
Tom

**Comment by JLetrange** [comment]
Hi Tom,

during our last tests we concluded together that the function _endsas worked as expected but an issue still remain on the disconnect function. I made some new tests based on the last changes done on the java side. I analyzed the code and provide some changes to generalize the way to disconnect. I will send you by email a reviewed version for the file saspy2j.java, in this version I added a method disconnect. Could you please take a look and provide your feedback on the changes brought. I tested on my side the following code and the jobs are properly deleted and created after a disconnect and a new sas submission:

```

import saspy
sas = saspy.SASsession(cfgname='winiomwin2')

print('Before Disconnect ...')
sas.submit('data a;x=1;run;','text')
print(sas.saslog())

sas.disconnect()
sas.disconnect() #called two times for testing purposes

print('After Disconnect ...')
sas.submit('data b;x=1;run;','text')
print(sas.saslog())

sas.disconnect()

print('After Disconnect ...')
sas.submit('data c;x=1;run;','text')
print(sas.saslog())

print('End SAS ...')
sas._endsas() 


```
Regards,
Joel.

**Comment by tomweber-sas** [comment]
We've been working on this issue offline. Going to summarize it here now so it's documented and others can find it here if they search for this. It turns out that, at their site, the object spawner is not reconnecting to the original workspace server after a disconnect. Instead, it is starting a new workspace server when the reconnect is requested. I've tried a number of different scenarios here w/ Grid and LSF and other IOM methods. In every case I can try, I always get reconnected to the original server, and never get a new server started. I suspect that there's a bug in the version of the object spawner they have, but they aren't in a position to try to diagnose this on their production system.

They also don't have any use for the disconnect/reconnect functionality, and simply want the workspace server to terminate when they are done. This does work with either an explicit endsas() call or implicitly when the python process terminates.

The outstanding problem however, is that as EG users try to use saspy, they may see the disconnect method and think that will terminate their workspace sever and use that method instead of endsas. If so, workspace servers will be left orphaned until their timeout period expires, and then they will be terminated. But if that's a hour, and there are many out there, that is not acceptable.

So, I've added a configuration definition key for the IOM access method 
```
            ‘reconnect’ : False,
```
When this is set to False (default it True; existing functionality), then a disconnect() invocation will not drive a disconnect from the server. This option is basically an override of the option in OMR that says whether or not reconnecting is allowed for that workspace server. With this set in the configuration definition, in the sascfg_personal.py file, users won't be able to accidentally use disconnect. This is acceptable since endsas is really what they would be wanting to do. In EG 'disconnect' is the term for terminating the workspace server; what endsas() is for saspy. EG doesn't appear to have the disconnect/reconnect ability, so there's not a corresponding term for that in EG. 

With this option set, all of the use cases behave as expected at the site. 

Tom
